### DISCUSSION TITLE
Question about comparison with Veros

### DISCUSSION THREAD

**Discussion by sophia-wright-blue** [CliMA/Oceananigans.jl#3128] [General]
thank you for the incredible effort to release and maintain this repo! I hope this is a fair request:

I have extensive background in Numerical solutions to PDEs, but am just beginning to get familiar with ocean circulation models.

it'd be very helpful if someone could help me compare the features and capability (current and planned) of **Oceananigans** (cool name btw!) vs **Veros** - https://github.com/team-ocean/veros, which is based on pyOM2 (https://wiki.cen.uni-hamburg.de/ifm/TO/pyOM2)

this will greatly help in understanding the 2 libraries and also which repo to focus on for different use cases.

thank you so much!

**Comment by glwagner** [comment]
Hi @sophia-wright-blue ! Do you mind if I convert this issue to a discussion?

I don't think any of the Oceananigans developers have much experience with Veros (perhaps understandably), but perhaps someone from our community will chime in about that. Possibly we can ask the Veros folks.

As for a summary of Oceananigans capabilities, your question is timely because we are planning to add a page to the documentation just for this purpose. I'll summarize the capabilities below --- but note that attempting to describe all the capabilities of an ocean model gets technical fairly quickly. So feel free to ask clarifying questions...

Oceananigans has three models:

1. `NonhydrostaticModel` which solves the non-hydrostatic Boussinesq equations with a rigid lid. The `NonhydrostaticModel` can only be used with a `RectilinearGrid`, which must have regular spacing in x, y and can be either regular or stretched in z. (There's an open PR to allow either x or y to be stretched as well, if z is regular.) This restriction has to do with the nonhydrostatic pressure solvers we support. Some important applications are:
* Direct numerical simulation using `CenteredSecondOrder` advection and `ScalarDiffusivity` with constant coefficients
* Large eddy simulation with `SmagorinskyLilly` or `AnisotropicMinimumDissipation` turbulence closures
* Idealized problems (eg two-dimensional turbulence) in 1-3 dimensions that use Laplacian, biharmonic, or implicit numerical dissipation.
2. `HydrostaticFreeSurfaceModel`, which solves the hydrostatic Boussinesq equations with linear free surface dynamics and a `z` coordinate. The hydrostatic model is useful for larger scale simulations, eg regional ocean modeling, and can be used with `RectilinearGrid` stretched in any or all directions as well as a `LatitudeLongitudeGrid`. We have a prototype closure for vertical mixing called `CATKEVerticalDiffusivity` (paper coming soon...) as well as a `RiBasedVerticalDiffusivity` and `ConvectiveAdjustment` that support implicit time discretization for vertical diffusion. We have a new high-order advection scheme specialized for geophysical turbulence / mesoscale turbulence problems that uses a vector invariant formulation. We can treat the free surface dynamics implicitly in time with various methods (FFT-based on rectilinear grids with flat bottoms, or preconditioned conjugate gradient based with various preconditioners in other situations). We also have a `SplitExplicitFreeSurface`.

All models support complex domains / bathymetry via `ImmersedBoundaryGrid`, a variety of Coriolis forces (`FPlane`, `BetaPlane`, as well as some more exotic options), a wide variety of boundary conditions for any prognostic variable on any model boundary. Note however that the pressure solver for `NonhydrostaticModel` is only approximate on `ImmersedBoundaryGrid` (we are working on an exact pressure solver for immersed boundaries).

We don't support simulations on the full sphere, or sea ice, yet, but we are working on both of those.

Let me know if you have any specific questions about capabilities, I'm curious to hear what problems you're interested in tackling.

**Comment by sophia-wright-blue** [comment]
thank you so much for the detailed response @glwagner - that's very helpful - it's a great idea to turn this into a discussion - I'll close this issue - very exited to learn more about Oceananigans (and Veros!)
